### PR TITLE
[ORCA] Avoid pushdown of predicate with set-returning function

### PR DESCRIPTION
**Pull request by dgkimura**
Issue is that if a predicate with a set-returning function is pushed
down then it can lead to bad execution because there exist cases where
the below context cannot accept a set. Logic already existed to prevent
such a case, but it wasn't sufficient.

Following example demonstrates such a case:
  ```sql
  CREATE TABLE t(a int[]) ;
  INSERT INTO t VALUES (ARRAY[1, 2]);

  SELECT * FROM (SELECT unnest(t1.a) a_unnest FROM t t1, t t2) s WHERE a_unnest IS NOT NULL;
  ```

### CODE REVIEW

**Review by gpopt — APPROVED** [review]
+1

**Comment by gpopt** [comment]
@dgkimura We need to update the plan part in the comment of `JoinOnViewWithSetReturningColumn.mdp`

**Comment by gpopt** [comment]
Oh, by the way, will Orca push it down if the project column contains subquery? We should disable it for this case.

**Comment by dgkimura** [comment]
On Thursday, September 29, 2022 2:18 PM gpopt wrote:
> Oh, by the way, will Orca push it down if the project column contains subquery? We should disable it for this case.

Yes, based on the code implementation of `DeriveHasNonScalarFunction()` I think it could.  I tried various queries, but I couldn't get it to push down because the plan shape didn't match SELECT over PROJECT over NARY-JOIN.  Do you have a specific problematic example in mind?

**Comment by gpopt** [comment]
```sql
 create table foo(a int, b int, c int);
 create table bar(a int, b int, c int);
```

```sql
explain select * from (select (foo.a in (select a from bar))::text as xx, foo.b from foo,bar) t  where xx='dd';
LOG:  statement: explain select * from (select (foo.a in (select a from bar))::text as xx, foo.b from foo,bar) t  where xx='dd';
LOG:  2022-09-29 17:02:38:142549 PDT,THD000,TRACE,"
Algebrized query:
+--CLogicalSelect
   |--CLogicalProject
   |  |--CLogicalNAryJoin
   |  |  |--CLogicalGet "foo" ("foo"), Columns: ["a" (0), "b" (1), "c" (2), "d" (3), "ctid" (4), "xmin" (5), "cmin" (6), "xmax" (7), "cmax" (8), "tableoid" (9), "gp_segment_id" (10)] Key sets: {[4,10]}
   |  |  |--CLogicalGet "bar" ("bar"), Columns: ["a" (11), "b" (12), "c" (13), "ctid" (14), "xmin" (15), "cmin" (16), "xmax" (17), "cmax" (18), "tableoid" (19), "gp_segment_id" (20)] Key sets: {[3,9]}
   |  |  +--CScalarConst (1)
   |  +--CScalarProjectList
   |     +--CScalarProjectElement "xx" (31)
   |        +--CScalarFunc (text)
   |           +--CScalarSubqueryAny(=)["a" (21)]
   |              |--CLogicalGet "bar" ("bar"), Columns: ["a" (21), "b" (22), "c" (23), "ctid" (24), "xmin" (25), "cmin" (26), "xmax" (27), "cmax" (28), "tableoid" (29), "gp_segment_id" (30)] Key sets: {[3,9]}
   |              +--CScalarIdent "a" (0)
   +--CScalarCmp (=)
      |--CScalarIdent "xx" (31)
      +--CScalarConst (797270572.000)

Algebrized preprocessed query:
+--CLogicalProject
   |--CLogicalNAryJoin
   |  |--CLogicalSelect
   |  |  |--CLogicalGet "foo" ("foo"), Columns: ["a" (0), "b" (1), "c" (2), "d" (3), "ctid" (4), "xmin" (5), "cmin" (6), "xmax" (7), "cmax" (8), "tableoid" (9), "gp_segment_id" (10)] Key sets: {[4,10]}
   |  |  +--CScalarCmp (=)
   |  |     |--CScalarFunc (text)
   |  |     |  +--CScalarSubqueryAny(=)["a" (21)]
   |  |     |     |--CLogicalGet "bar" ("bar"), Columns: ["a" (21), "b" (22), "c" (23), "ctid" (24), "xmin" (25), "cmin" (26), "xmax" (27), "cmax" (28), "tableoid" (29), "gp_segment_id" (30)] Key sets: {[3,9]}
   |  |     |     +--CScalarIdent "a" (0)
   |  |     +--CScalarConst (797270572.000)
   |  |--CLogicalGet "bar" ("bar"), Columns: ["a" (11), "b" (12), "c" (13), "ctid" (14), "xmin" (15), "cmin" (16), "xmax" (17), "cmax" (18), "tableoid" (19), "gp_segment_id" (20)] Key sets: {[3,9]}
   |  +--CScalarConst (1)
   +--CScalarProjectList
      +--CScalarProjectElement "xx" (31)
         +--CScalarFunc (text)
            +--CScalarSubqueryAny(=)["a" (21)]
               |--CLogicalGet "bar" ("bar"), Columns: ["a" (21), "b" (22), "c" (23), "ctid" (24), "xmin" (25), "cmin" (26), "xmax" (27), "cmax" (28), "tableoid" (29), "gp_segment_id" (30)] Key sets: {[3,9]}
               +--CScalarIdent "a" (0)
",
```
We pushed the subquery down.

We will see assertion error if the subquery is correlated subquery:
```sql
explain select * from (select (foo.a in (select a+foo.c from bar))::text as xx, foo.b from foo,bar) t  where xx='dd';
LOG:  statement: explain select * from (select (foo.a in (select a+foo.c from bar))::text as xx, foo.b from foo,bar) t  where xx='dd';
LOG:  2022-09-29 17:04:39:212310 PDT,THD000,ERROR,"CColumnFactory.cpp:375: Failed assertion: fres
Stack trace:
1    0x000000010190e8e0 gpos::CException::Raise + 112
2    0x0000000101b2e0fc gpopt::CColumnFactory::AddComputedToUsedColsMap + 424
3    0x0000000101b803c0 gpopt::CQueryContext::MapComputedToUsedCols + 212
4    0x0000000101b80420 gpopt::CQueryContext::MapComputedToUsedCols + 308
5    0x0000000101b80420 gpopt::CQueryContext::MapComputedToUsedCols + 308
6    0x0000000101b80420 gpopt::CQueryContext::MapComputedToUsedCols + 308
7    0x0000000101b80420 gpopt::CQueryContext::MapComputedToUsedCols + 308
8    0x0000000101b80420 gpopt::CQueryContext::MapComputedToUsedCols + 308
9    0x0000000101b800f8 gpopt::CQueryContext::CQueryContext + 888
10   0x0000000101b80498 gpopt::CQueryContext::CQueryContext + 88
11   0x0000000101b80c98 gpopt::CQueryContext::PqcGenerate + 1540
12   0x0000000101cdde08 gpopt::COptimizer::PdxlnOptimize + 1260
13   0x0000000101eb9064 COptTasks::OptimizeTask + 1780
14   0x00000001019293b4 gpos::CTask::Execute + 196
15   0x000000010192c768 gpos::CWorker::Execute + 208
16   0x0000000101927e24 gpos::CAutoTaskProxy::Execute + 336
17   0x000000010192f140 gpos_exec + 1104
```

I think we need to fix those 2 cases.

**Comment by dgkimura** [comment]
On Thursday, September 29, 2022 5:10 PM gpopt wrote:
> I think we need to fix those 2 cases.

Thanks Hansome for the example!!  I added commit [60f6855](https://github.com/greenplum-db/gpdb/pull/14201/commits/60f6855a62d027c3726c42cd5fe2f3d463205856) to fix along with a test case.

**Comment by gpopt** [comment]
Can we also add a non-correlated subquery (1st example) into the test case? It can also be in the ICW.

**Comment by dgkimura** [comment]
On Thursday, September 29, 2022 6:22 PM gpopt wrote:
> Can we also add a non-correlated subquery (1st example) into the test case? It can also be in the ICW.

I think we pick a correlated plan in both cases.  I see non-correlated alternatives, but I can't seem to force generating only non-correlated plans.  I think I want opposite of `optimizer_enforce_subplans`.  I'll look to see if I can tweak another GUC or add data to coerce picking a non-correlated plan.

**Comment by gpopt** [comment]
Hmm, I think the first query is non-correlated:
```sql
explain select * from (select (foo.a in (select a from bar))::text as xx, foo.b from foo,bar) t  where xx='dd';
```

**Review by gpopt — APPROVED** [review]
+1

File: src/backend/gporca/data/dxl/minidump/JoinOnViewWithSetReturningColumn.mdp [hunk 1]
```diff
@@ -318,7 +318,7 @@
     <dxl:Plan Id="0" SpaceSize="6">
       <dxl:GatherMotion InputSegments="0,1,2" OutputSegments="-1">
         <dxl:Properties>
-          <dxl:Cost StartupCost="0" TotalCost="862.000682" Rows="1.000000" Width="16"/>
+          <dxl:Cost StartupCost="0" TotalCost="862.000660" Rows="1.000000" Width="16"/>
```
Comment (chrishajas) [inline src/backend/gporca/data/dxl/minidump/JoinOnViewWithSetReturningColumn.mdp:321]: It looks like this test was supposed to catch this issue right? With this PR it's not pushed down, so now it matches the intent.

**Review by chrishajas — APPROVED** [review]
LGTM

File: src/backend/gporca/data/dxl/minidump/JoinOnViewWithSetReturningColumn.mdp (hunk 1 above)
Comment (dgkimura) [reply src/backend/gporca/data/dxl/minidump/JoinOnViewWithSetReturningColumn.mdp:321]: Yes, that's correct.  I'm not sure how this missight originally slipped through.

**Comment by dgkimura** [comment]
Merged. Thanks for the reviews!

**Pull request merged**